lib/bolt/application.rb

```diff
@@ -0,0 +1,612 @@
+        end
+
+        plan_params  = pal.get_plan_info(plan)['parameters']
+        target_param = plan_params.dig('targets', 'type') =~ /TargetSpec/
+        node_param   = plan_params.include?('nodes')
+
+        if node_param && target_param
+          msg = "Plan parameters include both 'nodes' and 'targets' with type 'TargetSpec', " \
+                "neither will populated with the value for --nodes or --targets."
+          Bolt::Logger.warn("nodes_targets_parameters", msg)
+        elsif node_param
+          params['nodes'] = targets.join(',')
+        elsif target_param
+          params['targets'] = targets.join(',')
+        end
+      end
+
+      plan_context = { plan_name: plan, params: params }
+
+      executor.start_plan(plan_context)
+      result = pal.run_plan(plan, params, executor, inventory, plugins.puppetdb_client)
+      executor.finish_plan(result)
+
+      result
+    end
+
+    # TODO: This should probably be separate #show_plan and #list_plans methods.
+    #
+    # Show plan information.
+    #
+    # @param plan [String, NilClass] The name of the plan to show.
+    # @param filter [String, NilClass] A string to filter plans by.
+    # @return [Hash]
+    #
+    def plan_show(plan, filter: nil)
+      if plan
+        pal.get_plan_info(plan)
+      else
+        plans = filter_content(pal.list_plans_with_cache(filter_content: true), filter)
+        { plans: plans, modulepath: pal.user_modulepath }
+      end
+    end
+
+    # Show available plugins.
+    #
+    # @return [Hash]
+    #
+    def plugin_show
+      { plugins: plugins.list_plugins, modulepath: pal.user_modulepath }
+    end
+
+    # Initialize the current directory as a Bolt project.
+    #
+    # @param name [String] The name of the project.
+    # @param [Bolt::Outputter] An outputter instance.
+    # @param modules [Array[String], NilClass] Modules to install.
+    # @return [Boolean]
+    #
+    def project_init(name, outputter, modules: nil)
+      Bolt::ProjectManager.new(config, outputter, pal)
+                          .create(Dir.pwd, name, modules)
+    end
+
+    # Migrate a project to current best practices.
+    #
+    # @param [Bolt::Outputter] An outputter instance.
+    # @return [Boolean]
+    #
+    def project_migrate(outputter)
+      Bolt::ProjectManager.new(config, outputter, pal).migrate
+    end
+
+    # Run a script on a list of targets.
+    #
+    # @param script [String] The path to the script to run.
+    # @param targets [Array[String]] The targets to run on.
+    # @param arguments [Array[String], NilClass] Arguments to pass to the script.
+    # @param env_vars [Hash] Environment variables to set on the target.
+    # @return [Bolt::ResultSet]
+    #
+    def script_run(script, targets, arguments: [], env_vars: {})
+      future = executor.future&.fetch('file_paths', false)
+      script = find_file(script, future)
+
+      Bolt::Util.validate_file('script', script)
+
+      with_benchmark do
```

Comment:
I just realized that previously we were doing all the file lookup and validation as part of the benchmark, and now it only wraps the actual script running. I personally don't particularly care one way or the other, but it'll be interesting to see if people notice Bolt *reporting* it ran a bit faster when it didn't...

Reply:
I feel like the real question is why we benchmark + report how quickly Bolt ran in the first place, and I think that it's mostly so people can verify and report issues when Bolt takes a long time without having to do their own benchmarking, which seems like a case for wrapping everything in a benchmark....but then again we only wrap commands that connect remotely in a benchmark, and finding the file definitely isn't part of that remote connection...Like, subscribing the outputted to the executor also used to be part of the benchmark and while that shouldn't take any noticeable amount of time it's also not meaningful to benchmark - basically there are some obvious cases, and I feel like file loading is kind of a gray area.

Reply:
Ya, my reasoning for wrapping just the executor portion is that it seems like the only part you'd really care to benchmark. File lookup and validation and subscribing to the executor should have a negligible impact on execution time, and we already were doing a fair amount of setup/validation in the CLI prior to reaching the benchmarking anyway. 🤷 Honestly I don't really feel strongly one way or the other.

Reply:
It doesn't seem to matter in the grand scheme, but for my 2 cents, I'd vote keeping it this way (wrapping just the executor). I'm sure users would adjust pretty easily and it seems like a good house-cleaning move.